Manifold/Term+DefinitionalEquality.swift

```diff
@@ -1,35 +1,33 @@
 //  Copyright © 2015 Rob Rix. All rights reserved.
 
 extension Term {
-	public static func equate(left: Term, _ right: Term, _ environment: [Name:Term]) -> Term? {
-		let recur: (Term, Term) -> Term? = {
-			equate($0, $1, environment)
-		}
+	public static func equate(left: Term, _ right: Term, _ environment: [Name:Term], var visited: Set<Term> = []) -> Term? {
+		if left == right { return right }
 
-		let normalize: Term -> Term = { term in
-			term.weakHeadNormalForm(environment, shouldRecur: false)
-		}
+		let (leftʹ, visitedLeft) = left.weakHeadNormalForm(environment, shouldRecur: false, visited: visited)
+		let (rightʹ, visitedRight) = right.weakHeadNormalForm(environment, shouldRecur: false, visited: visited)
+		visited.unionInPlace(visitedLeft)
+		visited.unionInPlace(visitedRight)
 
-		let left = normalize(left)
-		let right = normalize(right)
+		if leftʹ == rightʹ { return rightʹ }
 
-		if left == right { return right }
+		switch (leftʹ.out, rightʹ.out) {
+		case (.Implicit, _):
+			return rightʹ
+
+		case (_, .Implicit):
+			return leftʹ
```

Comment:
These two cases do not _really_ implement unification, since we only “unify” pairwise and not using constraints & a disjoint set. However, they’re an extremely effective start.